godot-core/src/builtin/callable.rs

```diff
@@ -0,0 +1,190 @@
+        }
+    }
+
+    /// Returns the object on which this callable is called.
+    ///
+    /// Returns `None` when this callable doesn't have any target to call a method on.
+    ///
+    /// _Godot equivalent: `get_object`_
+    pub fn object(&self) -> Option<Gd<Object>> {
+        // Increment refcount because we're getting a reference, and `InnerCallable::get_object` doesn't
+        // increment the refcount.
+        self.as_inner().get_object().map(|object| {
+            <<Object as GodotClass>::Mem as crate::obj::mem::Memory>::maybe_inc_ref(&object);
+            object
+        })
+    }
+
+    /// Returns the ID of this callable's object, see also [`Gd::instance_id`].
+    ///
+    /// Returns `None` when this callable doesn't have any target to call a method on.
+    ///
+    /// _Godot equivalent: `get_object_id`_
+    pub fn object_id(&self) -> Option<InstanceId> {
+        let id = self.as_inner().get_object_id();
+        InstanceId::try_from_i64(id)
+    }
+
+    /// Returns the 32-bit hash value of this callable's object.
+    ///
+    /// _Godot equivalent: `hash`_
+    pub fn hash(&self) -> u32 {
+        self.as_inner().hash().try_into().unwrap()
+    }
+
+    /// Returns true if this callable is a custom callable.
+    ///
+    /// Custom callables are mainly created from bind or unbind. In GDScript, lambda functions are also
+    /// custom callables.
+    ///
+    /// If a callable is not a custom callable, then it is considered a standard callable, this function is
+    /// the opposite of [`Callable.is_standard`].
+    ///
+    /// _Godot equivalent: `is_custom`_
+    ///
+    /// [`Callable.is_standard`]: https://docs.godotengine.org/en/stable/classes/class_callable.html#class-callable-method-is-standard
+    #[doc(alias = "is_standard")]
+    pub fn is_custom(&self) -> bool {
+        self.as_inner().is_custom()
+    }
+
+    /// Returns true if this callable has no target to call the method on.
+    ///
+    /// This is not the negated form of [`is_valid`], as `is_valid` will return `false` if the callable has a
+    /// target but the method does not exist.
+    ///
+    /// _Godot equivalent: `is_null`_
+    pub fn is_null(&self) -> bool {
+        self.as_inner().is_null()
+    }
+
+    /// Returns true if the callable's object exists and has a valid method name assigned, or is a custom
+    /// callable.
+    ///
+    /// _Godot equivalent: `is_valid`_
+    pub fn is_valid(&self) -> bool {
+        self.as_inner().is_valid()
+    }
+
+    #[doc(hidden)]
+    pub fn as_inner(&self) -> inner::InnerCallable {
+        inner::InnerCallable::from_outer(self)
+    }
+}
+
+impl_builtin_traits! {
+    for Callable {
+        Default => callable_construct_default;
+        // Equality depends on equality of the object/custom object of the callable's target.
+        // Which may be partial, so we cannot be sure that equality of Callable is total.
+        PartialEq => callable_operator_equal;
```

Comment:
Is it though?

We have a blanket
```rs
impl<T: GodotClass> Eq for Gd<T> {}

impl<T: GodotClass> PartialEq for Gd<T> {
    /// ⚠️ Returns whether two `Gd` pointers point to the same object.
    ///
    /// # Panics
    /// When `self` or `other` is dead.
    fn eq(&self, other: &Self) -> bool {
        // Panics when one is dead
        self.instance_id() == other.instance_id()
    }
}
```

So equality does not depend on contents of `T`.

Reply:
i think it is actually a total equality yeah, but custom callables in godot are free to implement equality however they want. from what i can tell it is always either just a pointer equality or object + method equality. but im not sure if godot guarantees that all custom callables in godot will be implemented to have a total equality.

Reply:
I see, but the current comment says it depends on the object (which doesn't make clear that they are free to implement however they want).

Reply:
fair, i wrote `object/custom object` where `custom object` was meant to refer to custom callable objects. but i can be more clear about that